libs/blocks/modal/modal.js

```diff
@@ -32,45 +32,64 @@ function closeModals(modals) {
       }
       modal.remove();
     });
-    window.history.pushState('', document.title, `${window.location.pathname}${window.location.search}`);
+    if(removeHash) {window.history.pushState('', document.title, `${window.location.pathname}${window.location.search}`);}
   }
 }
 
-export async function getModal(el) {
+function handleCustomModal(custom, dialog) {
+  dialog.id = custom.id;
+  dialog.classList.add(custom.class);
+  if (custom.closeEvent) {
+    dialog.addEventListener(custom.closeEvent, () => {
+      closeModals([dialog], false);
+    });
+  }
+  return custom.content;
+}
+
+async function handleAnchorModal(el, dialog) {
   const details = getDetails(el);
   if (!details) return null;
 
+  dialog.id = details.id;
+
+  const linkBlock = document.createElement('a');
+  linkBlock.href = details.path;
+
+  const { default: getFragment } = await import('../fragment/fragment.js');
+  await getFragment(linkBlock, dialog);
+
+  return linkBlock;
+}
+
+export async function getModal(el, custom) {
   const curtain = createTag('div', { class: 'modal-curtain is-open' });
   const close = createTag('button', { class: 'dialog-close', 'aria-label': 'Close' }, CLOSE_ICON);
   const dialog = document.createElement('div');
   dialog.className = 'dialog-modal';
-  dialog.id = details.id;
+
+  const content = custom ? handleCustomModal(custom, dialog) : await handleAnchorModal(el, dialog);
+  if (!content) return;
 
   close.addEventListener('click', (e) => {
-    closeModals([dialog]);
+    closeModals([dialog], !custom);
     e.preventDefault();
   });
 
   curtain.addEventListener('click', (e) => {
     // on click outside of modal
     if (e.target === curtain) {
-      closeModals([dialog]);
+      closeModals([dialog], !custom);
     }
   });
 
   dialog.addEventListener('keydown', (event) => {
     if (event.keyCode === 27) {
-      closeModals([dialog]);
+      closeModals([dialog], !custom);
     }
   });
 
-  const linkBlock = document.createElement('a');
-  linkBlock.href = details.path;
-
-  const { default: getFragment } = await import('../fragment/fragment.js');
-  await getFragment(linkBlock, dialog);
-
-  dialog.append(close, linkBlock);
+  dialog.append(close, content);
```

Comment:
Something about the above rubs me the wrong way and I cannot place my finger on it. Seems like we're passing around a lot of repetitive stuff to make our custom modal work, and I don't quite understand why. GeoRouting doesn't need to do anything when the close button is clicked. It should be agnostic to all of this code.

The whole conversion between passing in `!custom` which means `removeHash` seems weird to me. It would be nice if this was clearer what exactly is happening.

Reply:
Thx, I'll take a look at it. The code is necessary for when you've got a modal open and georouting appears on top, but yeah.. the way it's structured leaves a sour taste in the mouth doesn't it? I'll try to rewrite.